src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/IDE/VSSolutionManager.cs

```diff
@@ -205,6 +206,24 @@ private async Task InitializeAsync()
             var solutionLoadedGuid = VSConstants.UICONTEXT.SolutionExistsAndFullyLoaded_guid;
             _vsMonitorSelection.GetCmdUIContextCookie(ref solutionLoadedGuid, out _solutionLoadedUICookie);
 
+            if (ErrorHandler.Succeeded(_vsSolution.GetProperty((int)__VSPROPID.VSPROPID_IsSolutionOpen, out object isSoltuionOpen)))
+            {
+                _isSolutionOpen = (bool)isSoltuionOpen;
+            }
+            else
+            {
+                _isSolutionOpen = false;
+            }
+
+            if (ErrorHandler.Succeeded(_vsSolution.GetProperty((int)__VSPROPID.VSPROPID_SolutionDirectory, out object solutionDirectory)))
+            {
+                _solutionDirectory = (string)solutionDirectory;
+            }
+            else
+            {
+                _solutionDirectory = null;
+            }
+
```

Comment:
I see that we have exactly the same logic in `OnAfterOpenSolution` even handler, should we check for `_solutionDirectory != null` before invoking the `VSSolution.GetPropoerty` API here or in OnAfterOpenSolution event handler.

https://github.com/NuGet/NuGet.Client/pull/4879/files#diff-bed53a8b2e737e49ff6141c4279e525dfa2d100eabf6510ef6e0a1746de10505R1133-R1141

Reply:
No, the two can't overlap. This function is `InitiaizeAsync`, which is where the `OnAfterOpenSolution` event is hooked up. That callback can never fire until after `InitializeAsync` completes. And since `VsSolutionManager` is lazy-loaded, it means that a solution might already be open by the time that `InitializeAsync` runs (so `OnAfterOpenSolution was not called on solution open), therefore this code is initialzing it with the correct value.